pkg/chain/celo/celoutil/mining_waiter.go

```diff
@@ -98,51 +130,65 @@ func (mw MiningWaiter) ForceMining(
 		if err != nil {
 			logger.Infof(
 				"transaction [%v] not yet mined: [%v]",
-				transaction.Hash.TerminalString(),
+				transaction.Hash().TerminalString(),
 				err,
 			)
 		}
 
-		// transaction mined, we are good
+		// Transaction mined, we are good.
 		if receipt != nil {
 			logger.Infof(
 				"transaction [%v] mined with status [%v] at block [%v]",
-				transaction.Hash.TerminalString(),
+				transaction.Hash().TerminalString(),
 				receipt.Status,
 				receipt.BlockNumber,
 			)
 			return
 		}
 
-		// transaction not yet mined, if the previous gas price was the maximum
-		// one, we no longer resubmit
-		gasPrice := transaction.GasPrice
+		// Transaction not yet mined, if the previous gas price was the maximum
+		// one, we no longer resubmit.
+		gasPrice := transaction.GasPrice()
 		if gasPrice.Cmp(mw.maxGasPrice) == 0 {
-			logger.Infof("reached the maximum allowed gas price; stopping resubmissions")
+			logger.Infof(
+				"reached the maximum allowed gas price; " +
+					"stopping resubmissions",
+			)
 			return
 		}
 
-		// if we still have some margin, add 20% to the previous gas price
+		// If we still have some margin, add 20% to the previous gas price.
 		twentyPercent := new(big.Int).Div(gasPrice, big.NewInt(5))
 		gasPrice = new(big.Int).Add(gasPrice, twentyPercent)
 
-		// if we reached the maximum allowed gas price, submit one more time
-		// with the maximum
+		// If we reached the maximum allowed gas price, submit one more time
+		// with the maximum.
 		if gasPrice.Cmp(mw.maxGasPrice) > 0 {
 			gasPrice = mw.maxGasPrice
 		}
 
-		// transaction not yet mined and we are still under the maximum allowed
+		// Transaction not yet mined and we are still under the maximum allowed
 		// gas price; resubmitting transaction with 20% higher gas price
-		// evaluated earlier
+		// evaluated earlier.
 		logger.Infof(
-			"resubmitting previous transaction [%v] with a higher gas price [%v]",
-			transaction.Hash.TerminalString(),
+			"resubmitting previous transaction [%v] "+
+				"with a higher gas price [%v]",
+			transaction.Hash().TerminalString(),
 			gasPrice,
 		)
-		transaction, err = resubmitFn(gasPrice)
+
+		// Copy transactor options.
+		newTransactorOptions := new(bind.TransactOpts)
+		*newTransactorOptions = *originalTransactorOptions
+		newTransactorOptions.GasLimit = originalTransaction.Gas()
```

Comment:
We need to reassign. Suppose we don't set the `gasLimit` in the first `bind.TransactionOptions`. The `go-ethereum` code will estimate the limit and use it for the given transaction but it doesn't set the limit back to the `bind.TransactionOption`. That means the `gasLimit` will remain unset in the `bind.TransactionOption` during the resubmission and we always need to copy it from the original transaction.

Regarding the place where we should set it, I think that making it at the mining waiter level is the right choice. This is because the mining waiter is responsible for constructing the replacement transaction and increasing the mining change. The specific logic used by the mining waiter is an implementation detail. Today it bumps the `gasFeeCap` and `gasTipCap` but in the future, it might be something else. An alternative could be setting the `gasLimit` at the contract binding level but in that case, we set `bind.TransactionOpts` in multiple places which seems not elegant to me.

Reply:
> Suppose we don't set the `gasLimit` in the first `bind.TransactionOptions`. The `go-ethereum` code will estimate the limit and use it for the given transaction but it doesn't set the limit back to the `bind.TransactionOption`. That means the `gasLimit` will remain unset in the `bind.TransactionOption` during the resubmission (...)

Agreed, that's also my understanding. I think I do not agree about the last part though:

> (...) and we always need to copy it from the original transaction.

We need to copy it from the original transaction if we do not want `go-ethereum` to run the estimate again. If we do not copy it from the original transaction, then for each `ResubmitTransactionFn` call, `go-etherum` will run the estimate again. 

Do we agree on this piece?

Because if so, what I am saying is that we should *probably* allow `go-ethereum` to run the estimate again. In some cases, when the execution path in the contract does not change it might not be necessary. But in some other cases, I think it is necessary. Two examples:
- There is `if (tx.gasPrice > n)` condition in the code.
- Another transaction was mined before the one we are bumping the gas price for. As a result, the execution path in our transaction changed and now it needs more gas.

Reply:
Ok, got your point and I'm not against getting rid of the instructions which copy the gas limit. However, in case the client code sets their own `gasLimit` while invoking the given contract method, we need to preserve that. If we won't be copying the `gasLimit` from original tx, all resubmissions will be re-estimated instead of taking the value defined by the client. Of course, we can add a flag to the mining waiter (e.g. `reestimateGasLimit`) and set it according to initial options passed by the client. This will complicate the mining waiter behavior regarding the gas limit. Does this work for us?

Reply:
I'd say it should happen in `func(newTransactorOptions *bind.TransactOpts) (*types.Transaction, error) ` in the contract binding - if the original `transactorOptions` had `GasLimit` set, it should be rewritten to `newTransactorOptions` (without copying, we can just set this value). The mining waiter should not care about it, it's beyond its responsibility.

Reply:
See https://github.com/keep-network/keep-common/pull/84/commits/0a1c4f096d05b8347402b02d98c0dba6dd908f05